doc/source/train/user-guides/persistent-storage.rst

```diff
@@ -1,62 +1,387 @@
+--------------------------------------------
+
+.. tip::
+
+    Cloud storage is the recommended persistent storage option.
+
+If all nodes in a Ray cluster have access to cloud storage, then all outputs can be uploaded to a shared cloud bucket.
+
+Use cloud storage by specifying a bucket URI as the :class:`RunConfig(storage_path) <ray.train.RunConfig>`:
+
+.. code-block:: python
+
+    from ray import train
+    from ray.train.torch import TorchTrainer
+
+    trainer = TorchTrainer(
+        ...,
+        run_config=train.RunConfig(
+            storage_path="s3://bucket-name/sub-path/",
+            name="experiment_name",
+        )
+    )
+
+
+In this example, all files are uploaded to shared storage at ``s3://bucket-name/sub-path/experiment_name`` for further processing.
+
+
+Network filesystem (NFS)
+------------------------
+
+If all Ray nodes have access to a network filesystem, e.g. AWS EFS or Google Cloud Filestore,
+then all outputs can be saved to this shared filesystem.
+
+Use NFS by specifying the mount path as the :class:`RunConfig(storage_path) <ray.train.RunConfig>`:
+
+.. code-block:: python
+
+    from ray import train
+    from ray.train.torch import TorchTrainer
+
+    trainer = TorchTrainer(
+        ...,
+        run_config=train.RunConfig(
+            storage_path="/mnt/cluster_storage",
+            name="experiment_name",
+        )
+    )
+
+In this example, all files are saved to ``/mnt/cluster_storage/experiment_name`` for further processing.
+
+
+Local storage
+-------------
+
+Using local storage for a single-node cluster
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you're just running an experiment on a single node (e.g., on a laptop), Ray Train will use the
+local filesystem as the storage location for checkpoints and other artifacts.
+Results are saved to ``~/ray_results`` in a sub-directory with a unique auto-generated name by default,
+unless you customize this with ``storage_path`` and ``name`` in :class:`~ray.train.RunConfig`.
+
+
+.. code-block:: python
+
+    from ray import train
+    from ray.train.torch import TorchTrainer
+
+    trainer = TorchTrainer(
+        ...,
+        run_config=train.RunConfig(
+            storage_path="/tmp/custom/storage/path",
+            name="experiment_name",
+        )
+    )
+
+
+In this example, all experiment results can found locally at ``/tmp/custom/storage/path/experiment_name`` for further processing.
+
+
+Using local storage for a multi-node cluster
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. warning::
+
+    When running on multiple nodes, using the local filesystem of the head node as the persistent storage location is no longer supported.
+
+    If you save checkpoints with :meth:`ray.train.report(..., checkpoint=...) <ray.train.report>`
+    and run on a multi-node cluster, Ray Train will raise an error if NFS or cloud storage is not setup.
+    This is because Ray Train expects all workers to be able to write the checkpoint to
```

Comment:
I like the idea of `Ray Train expects all workers to be able to write the checkpoint to the same persistent storage location.`, that should be the core concept of persistent story. What about moving it to the very top of this user guide?

But logically using head node as persistent storage is also `the same persistent storage location`. The reason why we no longer support it is because of network communication bottleneck?

Reply:
yeah, the deprecation of head node as persistnet storage is described in the issue.